[5단계 미션] 임규영 제출합니다.

gxuoo opened the pull request:
안녕하세요 김의천 리뷰어님!
그리디 FE 2기 임규영입니다. 1주차 미션 이후로 다시 인사드리네요~!! 잘 부탁드리겠습니다!!

이번 미션은 json-server를 활용하여 db.json의 데이터에 접근하여 GET 하여 정보를 보여준 후, 추가된 정보에 대해 POST하여 저장하는 미션이었습니다. 미션 진행은 가이드에 나와있는 GET, POST 예시 코드를 참고하여 진행하였습니다.

## 컴포넌트 설명
### App
1. `<header>` 태그에 들어갈 내용인 Header
2. 레스토랑 카테고리를 필터링 해주는 RestaurantFilter
3. 필터링된 레스토랑 리스트를 보여주는 RestaurantList
4. [Add, Detail] Modal의 Open & Close 상태 관리 및 각 Modal 컴포넌트
이렇게 총 4가지를 포함하고 있습니다. 

### Header
- 이 페이지의 제목인 '점심 뭐 먹지'를 보여주고, 음식점 추가 버튼을 클릭하면 `AddModal`이 나타나도록 설정합니다.

### RestaurantContext
- Step4를 진행하면서 과도한 props Drilling을 방지하기 위한 Context API를 도입을 고려해보라는 리뷰를 받고 자주 사용되는 상태들을 `RestaurantProvider`로 담아서 관리하고 있습니다.
- 이번 미션에서는 `getRestaurants`를 추가하여 레스토랑 리스트를 리랜더링할 때 사용할 수 있도록 해당 함수를 추가하였습니다.

### RestaurantFilter
- 카테고리 선택을 담당하는 컴포넌트로, `setSelectedCategory`를 통해 선택된 카테고리를 넘겨줍니다.
- `FilteredCategoryOptions`은 카테고리가 저장된 상수에서 선택된 카테고리를 반환하고, `CategoryOption` 각 카테고리를 <option> 태그로 나타낸 컴포넌트입니다.

### RestaurantList
- 선택된 카테고리에 따른 레스토랑 리스트를 만들어 보여주는 역할을 합니다.
- `RestaurantListItem`컴포넌트는 해당 레스토랑을 클릭 시, `setRestaurantItem`으로 해당 레스토랑을 설정한 뒤에 DetailModal을 Open 하도록 설정합니다.

### RestaurantDetailModal
- `restaurantItem` 값을 가지고 해당 레스토랑의 이름, 정보를 보여줍니다. backdrop이나 닫기 버튼을 클릭하면 Modal이 사라집니다.

### RestaurantAddModal
- Header의 추가 버튼을 클릭하면 Open 됩니다.
- 음식점 카테고리, 이름, 정보를 입력한 뒤 추가하기 버튼을 누르면 db.json에 입력한 레스토랑 정보가 추가되고, Modal이 사라집니다.

----

## 🗂️ 프로젝트 구조 (Directory Structure)
<img width="995" alt="Step5 Structure" src="https://github.com/user-attachments/assets/3d34850c-eb63-462e-abf2-503884d38283" />

## 🎆 실행 영상
https://github.com/user-attachments/assets/23a77aac-6801-42da-a59b-eace92dc5c72

## 😮 구현하면서 궁금했던 점
- 이전 4단계 미션에 대한 리뷰를 받고 나서, Context API를 사용하여 전역적으로 상태를 관리했습니다. 
- 초기에는 `getRestaurants`의 위치가 `RestaurantContext` 컴포넌트에 없었고 `RestaurantList`에 위치했었습니다. 그리고 `RestaurantAddModal`에서 `handleSubmit` 함수 내부에 레스토랑 정보를 추가하는 로직을 만든 뒤, `RestaurantList`랜더링을 위해 `getRestaurants`의 위치를 `RestaurantContext` 컴포넌트 내부로 옮겼는데 해당 방식에서 발생할 수 있는 문제점이 있는지 궁금합니다.

```javascript
// RestaurantContext 컴포넌트 중 일부
const getRestaurants = async () => {
    try {
        const response = await fetch("http://localhost:3000/restaurants");
        const data = await response.json();
        setRestaurants(data);
    } catch (err) {
        console.error("레스토랑 데이터를 불러오는 데 실패했습니다:", err);
    }
}

useEffect(() => {
    getRestaurants();
}, []);
``` 
```javascript
// RestaurantAddModal 컴포넌트의 handleSubmit 함수의 일부
try {
    const response = await fetch("http://localhost:3000/restaurants", {
        method: "POST",
        headers: {
            "Content-Type": "application/json",
        },
        body: JSON.stringify(newRestaurant),
    });

    if (!response.ok) {
        throw new Error("서버 응답 오류");
    }
} catch (error) {
    console.error("실패:", error);
}

getRestaurants();  // RestaurantList 리랜더링을 위한 getRestaurants 함수 호출
```

gxuoo committed:
feat: step1 컴포넌트 기초 틀 구현

가장 큰 Container라는 틀을 통해 Head, Main, Aside 3가지로 구분 지었고, 각각의 해당되는 css 파일을 구분했습니다.

gxuoo committed:
feat: Header 컴포넌트 구현

HeadContainer에 들어가는 Header 컴포넌트 구현

gxuoo committed:
feat: 레스토랑 필터, 리스트, 값 컴포넌트 구현

MainContainer에 들어갈 레스토랑 카테고리에 따른 필터와 레스토랑 리스트, 리스트에 들어갈 각각의 레스토랑 값에 대한 컴포넌트를 생성했습니다. 그리고 이에 따른 png 파일의 저장 위치를 React 폴더 구조에 알맞게 변경해보았습니다.

gxuoo committed:
feat: Aside 컴포넌트 구현

레스토랑의 정보를 띄워주는 컴포넌트, 레스토랑을 추가할 수 있는 컴포넌트를 구현했습니다.

gxuoo committed:
fix: 파일명 수정

gxuoo committed:
refactor: 파일명 수정, .png 위치 수정

gxuoo committed:
feat: module.css 적용 및 카테고리 설정에 따른 필터된 음식점 목록 출력

전역 App.css 와 Head, Main, Aside 각 컴포넌트에서 사용되는 css에 대해 코드를 수정하였고, 카테고리 설정에 따른 필터링된 음식점 목록을 보여주는 기능을 추가했습니다.

gxuoo committed:
fix: 리뷰 반영 수정

네이밍 변경, 불필요한 Fragment 제거, 카테고리 컴포넌트 추가, map 함수 key 변경

gxuoo committed:
Merge branch 'gxuoo' into step2

gxuoo committed:
fix: 2차 리뷰 반영 수정

불필요한 Container 컴포넌트 제거, 컴포넌트 외부에 상수 선언, 네이밍 변경

gxuoo committed:
refactor: 2단계 리뷰 반영 리팩토링

사용하지 않는 파일들 삭제

gxuoo committed:
feat: RestaurantDetailModal 구현

레스토랑을 클릭하면 상세 정보를 띄워주는 Modal 구현

gxuoo committed:
Merge remote-tracking branch 'upstream/gxuoo' into step3

gxuoo committed:
refacator: 사용하지 않는 파일 삭제

gxuoo committed:
feat: RestaurantAddModal 구현

레스토랑 추가하는 기능을 구현하였고 재사용 가능한 Modal 컴포넌트를 사용하여 AddModal, DetailModal에 적용시켰습니다.

gxuoo committed:
refactor: 이미지 파일 public 폴더로 이동

import 하는 방식이 아닌 public 폴더에서 정적 URL로 접근하여 이미지 Loading

gxuoo committed:
refactor: 변수 네이밍 변경

모달 상태 관리 변수 modalState를 isModalOpen 이라는 이름으로 rename 했습니다.

또한 Restaurant의 정보를 보여주는 컴포넌트인 RestaurantValue를 RestaurantListItem 으로 변경했습니다.

gxuoo committed:
refactor: module.css 개선 및 스타일 통일

module.css를 컴포넌트 별로 적용 시켰습니다. 또한 스타일이 일치되지 않았던 코드를 하나의 스타일로 일치시켰습니다.

gxuoo committed:
refactor: 이미지 파일 경로 설정

gxuoo committed:
fix: 카테고리 검증 로직 변경

기존의 alt 속성으로 비교 연산을 진행했던 것을 category key를 추가로 만들어 해당 key로 비교 연산을 진행합니다.

gxuoo committed:
refactor: 상수 네이밍 변경

upper case에 맞춰서 변경했습니다.

gxuoo committed:
refactor: 컴포넌트 네이밍 변경 및 Context API 도입

상태 관리를 위한 RestaurantContext를 추가하였고 각 컴포넌트의 역할과 기능에 따른 네이밍을 수정하였습니다.

gxuoo committed:
Merge branch 'gxuoo' into step4

gxuoo committed:
refactor: 현재 로직에 알맞은 레스토랑 멤버 정보 수정

기존에 제공된 파일과 현재 레스토랑 목록을 보여주고, 필터링하는 로직이 알맞지 않아 형식에 맞게 데이터를 수정했습니다.

gxuoo committed:
fix: step4 변경 내용 누락

step4에서 제거했어야할 상태를 삭제했습니다.

gxuoo committed:
feat: fetch API로 레스토랑 정보 불러오기

fetch API를 이용하여 db.json에 저장되어 있는 레스토랑 정보를 불러오게 했습니다.

gxuoo committed:
feat: fetch API를 사용하여 레스토랑 추가 구현

기존의 레스토랑을 추가하는 로직에 변경사항이 있습니다.

1. db.json 에 저장된 객체 형식에 맞게 변경
2. handleSubmit 함수를 비동기 함수로 변경
3. getRestaurants를 useEffect 밖에 선언하여 context value로 전달, AddModal이 닫혔을 때 레스토랑 목록이 리랜더링 될 수 있도록 변경

이렇게 총 3가지 변경 사항이 있습니다.

gxuoo committed:
refactor: 네이밍 변경

사용하지 않는 파일 삭제 및 네이밍 변경

gxuoo committed:
Merge branch 'gxuoo' into step5

wzrabbit commented on src/context/RestaurantContext.jsx:
```diff
@@ -22,13 +21,28 @@ export function RestaurantProvider({ children }) {
+            const data = await response.json();
+            setRestaurants(data);
+        } catch (err) {
+            console.error("레스토랑 데이터를 불러오는 데 실패했습니다:", err);
```
이 오류 메시지를 볼 독자를 누구로 잡으시고 이 코드를 작성해주셨는지 말씀해 주실 수 있나요?

gxuoo replied:
제가 콘솔에서 확인하는 용도로 작성한 오류 메시지 입니다.

wzrabbit replied:
그렇다면 어울리는 의도네요 😊

만약 사용자를 독자로 두시려고 말씀하셨다면 `console.log`는 사용자의 눈에 잘 띄지 않음을 이야기하며 개선을 요청드리려고 했습니다.
예외 처리까지 하는 것은 이 미션의 범위는 아니지만, 예외 처리 역시 애플리케이션을 만드는 데에 있어 중요한 요소이니 이후의 미션에서는 시간이 남으셨다는 가정 하에 예외 처리도 적용해 보시면 좋은 연습이 될 수 있을 거라 생각합니다

gxuoo committed:
refactor: merge 실수로 인한 파일 수정

컴포넌트 네이밍, css 수정이 있습니다.

wzrabbit commented on src/components/aside/RestaurantAddModal.jsx:
```diff
@@ -15,31 +15,47 @@ const CATEGORYMAP = {
-        const id = restaurants.length + 1;
+        const id = (restaurants.length + 1).toString().padStart(2, "0");
         const category = e.target.category.value;
         const alt = CATEGORYMAP[category];
```
`"CATEGORYMAP"` 은 제가 봤을 때는 두 개의 단어로 보이는데, 어떻게 생각하시나요? 단어가 여러 개고, 대문자 snake case를 사용 중이실 경우 `_`로 단어를 구분해야 하는 규칙이 있습니다. `CATEGORY_MAP`

gxuoo replied:
아 제가 한 단어로 착각하고 있었네요... 수정하겠습니다!

wzrabbit commented on src/context/RestaurantContext.jsx:
```diff
@@ -22,13 +21,28 @@ export function RestaurantProvider({ children }) {
+        }
+    }
+
+    useEffect(() => {
+        getRestaurants();
+    }, []);
```
이 코드와의 차이점은 무엇일까요?
```js
useEffect(() => {
  getRestaurants();
});
```

gxuoo replied:
의천님이 작성해주신 코드는 의존성 배열이 없는 코드로 RestaurantContext 컴포넌트가 랜더링될 때마다 getRestaurants() 함수가 호출되고, 의존성 배열로 빈 배열을 주입하게 되면 컴포넌트가 처음 랜더링될 때만 getRestaurants() 함수가 호출됩니다.

wzrabbit commented on src/components/aside/RestaurantAddModal.jsx:
```diff
@@ -15,31 +15,47 @@ const CATEGORYMAP = {
+            console.error("실패:", error);
+        }
+
+        getRestaurants();
```
오 좋아요 👍🏻 요구사항을 빠뜨리지 않고 잘 이행하셨군요

> 레스토랑 추가 모달에서 추가하기 버튼을 클릭하면 POST 요청을 보냅니다. 모달이 닫히고, **레스토랑 목록을 다시 불러옵니다.**

여러 개의 창을 켜고 테스트해봤는데 특정 창에서 추가한 음식점이 다른 창에서도 잘 반영되는 것을 확인했습니다

gxuoo replied:
감사합니다!

wzrabbit commented on src/components/aside/RestaurantAddModal.jsx:
```diff
@@ -15,31 +15,47 @@ const CATEGORYMAP = {
+        }
+
+        getRestaurants();
         setActiveModalIndex(0);
```
음, `setActiveModalIndex(0);` 만으로는 무슨 모달이 보여진다는 건지를 알아채기 어려웠던 것 같습니다. 확인해보았는데 `isModalOpen`, `activeModalIndex` 두 개의 상태를 이용해 모달을 관리하고 계시더라고요.

두 가지 아이디어를 한 번 추천드려볼래요!

### 1️⃣ 활성화되어 있는 모달을 문자열 상태로 관리하기
- `activeModalName` 과 같은 상태를 만들고, `"restaurantAddModal"`, `"restaurantDetailModal"`, `null` 등과 같은 값을 넣을 수 있도록 한다면 이름이 문자열에 드러나서 좋을 것 같아보여요. 상태도 하나로 줄어들어요.
- 보너스: 1️⃣ 번 아이디어를 채택할 경우의 한계는 활성화되어 있는 모달을 하나로만 지정할 수 있다는 거에요. 만약 여러 개의 모달도 활성화될 수 있도록 각각의 모달의 상태를 관리하려면 어떻게 해 볼 수 있을까요?

### 2️⃣ 지금의 구조를 유지하되, 매직 넘버를 지양하는 방향으로 관리하기
- 지금의 구조가 마음에 드신다면, `0`, `1`과 같은 의미를 알아보기 힘든 매직 넘버를 상수로 관리해 보시는 것은 어떨까요?

gxuoo replied:
저는 1번으로 하는 게 좋아보입니다! 이번에 축제 준비하면서 만들었던 미니게임에서도 모달을 switching 하면서 문자열 상태로 관리하는게 직관적으로 알아차릴 수 있었습니다. 문자열 상태로 관리하도록 수정하겠습니다.

wzrabbit replied:
좋아요! 요구사항에 따라 해볼만한 방법들도 달라질 거라 생각해요. 새로운 요구사항이 주어지면 그때는 규영님이 잘 해주실 거라 믿어요

wzrabbit commented on src/context/RestaurantContext.jsx:
```diff
@@ -22,13 +21,28 @@ export function RestaurantProvider({ children }) {
 
+    const getRestaurants = async () => {
+        try {
+            const response = await fetch("http://localhost:3000/restaurants");
+            const data = await response.json();
```
`await` 이 이 코드에서 필요한 이유가 무엇일까요? 얘가 무슨 역할을 하길래요?

gxuoo replied:
await은 비동기 작업들 간의 순서가 있다면, 해당 키워드를 통해 동기적으로 작업을 처리할 수 있습니다. await 키워드는 이전 응답이 완료될 때까지 기다렸다가 그 결과를 변수에 할당하는 것으로 이해하고 있습니다.

wzrabbit replied:
맞아요! `fetch`와 `response.json` 모두 비동기로 작동하는 함수들인데, 놓쳐서는 안 되는 점은 **두 함수는 모두 비동기로 작동함에도 어느 한쪽의 함수가 먼저 실행되어 결과를 전달**해야 한다는 점이에요. 함수를 async 함수로 만들고, 내부에 await 키워드를 사용하면, 비동기 작업이 실행되더라도 바로 다음 줄의 코드가 실행되지 않고 비동기 작업이 끝나 결과를 반환할 때까지 대기하게 됩니다. 잘 짚어주셨어요.

wzrabbit commented on src/components/aside/RestaurantAddModal.jsx:
```diff
@@ -15,31 +15,47 @@ const CATEGORYMAP = {
         e.preventDefault();
 
-        const id = restaurants.length + 1;
+        const id = (restaurants.length + 1).toString().padStart(2, "0");
```
상황에 따라 상당히 위험한 아이디 부여 방식이 될 수도 있을 것 같습니다. `restaurants.length`는 여러 개의 창에서 서로 실시간으로 항시 공유되는 값이 아닙니다. `restaurant.length`가 같은 상태로, 두 창에서 각각 음식점을 추가한다면 같은 아이디의 음식점이 추가될 수도 있을 것 같습니다. 어떻게 이 문제를 해결해 볼 수 있을까요?

gxuoo replied:
아하 생각해보니 그런 문제가 있네요... 이전 3단계? 미션을 진행하면서 id 값을 Date.now() 값을 부여했던 것으로 기억하는데, 이 값도 무조건 unique 한 값은 아니기에 해당 값에 `Date.now + '-' + 5자리 숫자` 형식으로 만들어서 저장하는 것으로 대체하면 해당 문제를 그나마 해결할 수 있다고 생각됩니다.

wzrabbit replied:
말씀하신 대로 하시면 제가 말씀드린 문제는 해결되고, 키가 충돌할 확률도 낮아지지만 마냥 안전하다고 보기는 어려울 수도 있을 것 같습니다. 이 미션에서는 사용자가 버튼을 눌러 키를 생성하는지라 문제는 거의 없을듯합니다. 하지만 리스트를 렌더링할 때 키 값들을 부여하는 상황 등 짧은 시간 안에 많은 개수의 키가 할당되는 상황이라면 위험할 수 있어보입니다.

- `Date.now()`의 경우 키가 충돌하려면 최소 단위인 1ms 안에 키가 두 번 이상 생성되어야 합니다. 1ms는 사람에게는 극단적으로 짧은 시간이지만 컴퓨터에게는 그렇지 않습니다. 1ms 안에 꽤 많은 키가 할당될 수도 있으며 이 경우에는 충돌이 일어납니다.
- `Math.random()`의 경우 값이 손실될 정도로 큰 값이 아니라면 그럭저럭 랜덤 키를 생성하는 역할은 할 수 있지만, 여전히 충돌 확률이 꽤 높은 것이 단점입니다. 우연히 두 개의 값을 뽑았을 때 두 값이 같을 확률은 굉장히 낮지만, 많은 개수의 키를 한꺼번에 생성하는 상황이면 의외로 충돌 확률이 크게 올라갑니다. 수의 범위가 100,000이라는 가정하에 키 값을 1,000개만 생성해도 충돌 확률은 95% 이상이 됩니다.

둘은 같이 사용되므로 더 충돌 확률이 내려가고 서로의 단점을 보완하지만, 여전히 아래의 상황에서는 위험할 수 있어보입니다.
- **짧은 시간 안에 많은 개수의 키가 할당되는 상황**: `Date.now()`가 우선 뚫리고, `Math.random()`에만 의존해야 하는 상황이 됩니다.

저는 개인적으로 `crypto.randomUUID()`를 추천드립니다. 라이브러리 설치가 필요 없으며 충돌 확률이 극단적으로 낮은 키를 생성합니다. 어느 정도냐면, 키 값을 100조 개 생성해도 키가 충돌할 확률이 복권 당첨 확률보다 훨씬 낮습니다. 그 외에 uuid를 생성하는 라이브러리 및 기술도 좋아보입니다.

```
>  crypto.randomUUID()
<- 'a3744bc8-f6cd-4508-ad83-b8d3bfacd82d'
```

wzrabbit reviewed (changes requested):
@gxuoo 👋🏻 

오랜만입니다, 잘 지내셨나요? 숫자 야구 미션의 처음과 self-paced-react 미션의 마지막을 제가 담당하는군요!

<img src="https://github.com/user-attachments/assets/31c41871-417d-440a-8a98-db19e91ffebc" width="100px" />

## 1️⃣ 질문/답변
> 초기에는 getRestaurants의 위치가 RestaurantContext 컴포넌트에 없었고 RestaurantList에 위치했었습니다. 그리고 RestaurantAddModal에서 handleSubmit 함수 내부에 레스토랑 정보를 추가하는 로직을 만든 뒤, RestaurantList랜더링을 위해 getRestaurants의 위치를 RestaurantContext 컴포넌트 내부로 옮겼는데 해당 방식에서 발생할 수 있는 문제점이 있는지 궁금합니다.

코드를 확인해 보았는데 전체적으로 비동기 요청 로직이 `RestaurantProvider`로 옮겨진 듯합니다. 다른 컴포넌트에서는 Provider가 제공하는 함수들을 사용만 하고 있고요. 음식점 리스트 값은 이 Provider와 크게 연관되어 있기에 Provider에 비동기 로직을 붙이는 것은 저에게는 어색하게 느껴지지 않고, 기술적으로 발생할 문제점도 저에게는 잘 보이지 않습니다. 혹시 문제점을 발견하셨거나 의심되는 정황이 있다면 저에게 공유해 주실 수 있나요?

## 2️⃣ Context API에 대한 이야기
**Context API**를 사용해 주셨더라고요. Context API나 그에 준하는 상태 관리 라이브러리를 접하는 미션은 더 뒤에 있을 것이기 때문에 그 미션에서 더 고민해 보실 것 같습니다. 이전 리뷰어님의 조언을 받아들이며 도입하셨음을 보았기 때문에 부연설명 겸 고민해 보실 만한 사항을 적어보겠습니다.

먼저 요약입니다.

1. depth가 깊어질수록 관리가 어려워지고, Context API를 도입해 볼 만하지만 depth가 깊다고 그 구조가 나쁜 구조라는 것은 아닙니다. 더 많은 요소들을 고려해야 합니다. 사실 **2 depth**는 얕은 편에 속합니다. **규영님이 도입하시면서 편리함을 느끼셨는지가 중요하다고 생각**합니다.
2. Context API를 도입할 경우 컴포넌트의 재사용성 등 props drilling이 아닌 다른 요소에도 영향을 줍니다.

아래의 내용은 좀 더 자세한 설명으로, 이해가 안 되신다면 요약만 읽고 가셔도 충분하다고 생각합니다. 어쨌든 Context API는 본 미션의 주제가 아니며 useEffect에 대해 고민해 보시는 것을 더 추천드리고 싶습니다.

<details>
<summary>depth가 높은 구조는 항상 나쁜 구조일까요</summary>
 <br>
<ul>
  <li>이전 리뷰어님이 말씀해 주신대로 과도한 props drilling은 유지보수를 어렵게 하는 것은 맞습니다. 그렇지만 depth가 깊다고 항상 이 구조는 나쁜 구조라 말할 수 있을까요? 관심사에서 벗어나지 않고 체계적으로 전달되는 데이터 흐름이라면 depth가 다소 깊더라도 props drilling을 그대로 두는 것이 좋은 방법이 될 수도 있습니다. 반면 depth가 얕은 편이더라도 이 상태를 쓰는 컴포넌트들이 여러 계층으로 흩어져있어 데이터를 운반하기에 드는 공수가 크다면 Context API의 도입은 좋은 방법이 될 수 있겠죠.</li>
<li>props와 state를 이용한 상태관리는 Context API 및 다른 상태 관련 라이브러리의 하위 호환이 아닙니다. props/state가 적합한 상황도 있고, 둘 모두 쓰는 경우도 있습니다.</li>
<li>결론적으로 Context API를 적절한 상황에 도입했다면 구조의 복잡함을 해소시킬 수 있지만, 잘못 사용하면 오히려 구조가 복잡해질 수도 있습니다.</li>
</ul>

여담으로 이전 단계에서 불필요한 Wrapper를 제거해 주시면서 복잡성을 줄이신 것은 좋다고 생각합니다.
</details>


<details>
<summary>Context API는 props drilling만 변화시키는 것이 아닙니다</summary>
<ul>
<br>
 <li>예를 들면 컴포넌트의 재사용성에 변화를 줍니다.</li>
  <li>props를 통해 데이터를 부여받는 컴포넌트는 props가 무엇인지만 중요하므로 서로 다른 기능을 담당하는 부모로 교체되어도 상관없습니다. 자식 컴포넌트는 자기 일을 그대로 하면 됩니다.</li>
<li>그러나, Context API의 상태를 사용하는 컴포넌트는 어떤 상태를 구독(Subscribe)하는지가 정해져 있기 때문에, 사용처가 특정 기능만을 위한 컴포넌트로 바뀝니다. 이것이 나쁘다는 것을 의미하지는 않지만, 광범위하게 재사용되는 공통 컴포넌트라면 재사용성을 상실할 수도 있게 됩니다.</li>
<li>결론적으로 도입되는 컴포넌트의 재사용성에 따라 좋은 설계가 될 수도 있고 나쁜 설계가 될 수도 있습니다. 도입 전 컴포넌트의 성격을 파악해보세요.</li>
</ul>
</details>

## 3️⃣ useEffect에 대한 이야기
이 미션의 중요 주제죠, useEffect? 상당히 특이한 사용법을 지니고 있으면서 잘못 사용하기 쉬운 훅이기도 합니다. 코드를 보았을 때는 전반적으로 잘 적용해 주신 것 같아, 저는 아래의 질문들을 드리는 것으로 `useEffect` 에 대해 탐구하시고 필요성을 생각해 보실 시간을 드리고 싶습니다.

- Q1. **사이드 이펙트**는 무엇일까요?
- Q2. 함수 및 컴포넌트가 **순수하다** 는 것은 무엇일까요?
- Q3. React는 왜 순수한 도메인을 지향할까요?
- Q4. `useEffect` 내부에 사이드 이펙트를 다루는 도메인들을 넣는 것으로 렌더링과 사이드 이펙트와 관련된 로직을 분리할 수 있습니다. 분리하는 이유가 무엇일까요?

나머지 리뷰들은 모두 코멘트에 적어보았습니다. 빈틈이 없어보여 저도 나름의 연구를 조금 해 보았습니다. 제 코멘트가 도움이 되었으면 합니다. 아 그리고 미션 문서를 봤는데 첨부되어 있는 글들이 꿀이더라고요, 꼭 한 번 확인해 보시면 좋겠어요

그럼, 규영님의 의견을 들려주세요!

gxuoo committed:
refactor: 상수를 대문자 snake case로 통일

상수 값은 대문자 snake case로 통일했습니다.

gxuoo committed:
refactor: Modal의 상태 관리 문자열로 변경

1. boolean 상태 변수로 Modal의 open & close 여부
2. 정수형 상태 변수로 Modal의 인덱스를 관리하였습니다.

이를 하나의 문자열 상태 변수로 관리하도록 변경했습니다.

gxuoo committed:
fix: 레스토랑 id 생성 로직 변경

Date.now + ' - ' + 5자리 랜덤 숫자의 조합인 문자열로 id 값을 변경하여 unique 하도록 변경하였습니다.

gxuoo reviewed (commented):
## 1️⃣에 대한 답변
따로 문제점을 발견하거나 의심되는 정황은 없었습니다!

## 2️⃣ Context API에 대한 답변
의천님이 적어주신 Context API에 대한 이야기를 모두 읽어봤습니다. 그 중 제일 인상이 깊고 다시 생각하게 됐던 점은 컴포넌트의 재사용성이었습니다.

아직까지 정확하게 컴포넌트의 재사용성에 대한 개념이 잘 와닿지 않는 것 같습니다. 모든 컴포넌트가 재사용성을 기반으로 구현해야 하는 것인지, 아니면 재사용 가능성이 높다고 기대되는 컴포넌트에 대하여 구현해야 하는 것인지 기준을 잡는다(?)는 게 아직 쉽지 않다고 생각합니다.

따라서, 이번 Context API를 사용하면서 props drilling을 방지하고, 이번 코드에서는 코드 가독성을 높일 수 있어서 좋다고는 생각했지만, 한편으로는 API를 사용하면서 발생할 수 있는 다른 문제점은 어떤 문제가 발생할 수 있을까? 라는 생각을 할 수 있었던 계기가 되었던 것 같습니다.

## 3️⃣ useEffect에 대한 답변
### A1. 사이드 이펙트
- 컴포넌트의 렌더링 이외의 다른 이벤트 리스너로 발생하는 이펙트를 말하는 것으로 알고 있습니다.

### A2. 함수 및 컴포넌트가 순수하다는 것?
- 렌더링 전에 존재했던 객체나 변수를 변경하지 않아야 합니다.
- 같은 입력, 같은 출력. 입력이 같을 경우, 컴포넌트는 항상 같은 JSX를 반환해야 합니다.

### A3. React는 왜 순수한 도메인을 지향할까?
- 컴포넌트나 함수를 순수하게 작성한다면, 예측 가능성이 올라가면서 랜더링 결과를 예상하기 쉽습니다. 이에 따라 디버깅 및 테스트를 쉽게 할 수 있습니다.

### A4. 렌더링과 사이드 이펙트를 분리하는 이유?
- 렌더링 후에 사이드 이펙트가 일어나야 컴포넌트의 순수함을 지킬 수 있기 때문입니다. 그리고 불필요한 사이드 이펙트의 호출을 막을 수 있다고 생각합니다.

----
이상으로 5단계 미션 리뷰 반영을 마치겠습니다. 감사합니다!

wzrabbit reviewed (approved):
@gxuoo 👋🏻 

바쁘신 와중에도 미션 수행하시느라 고생 많으셨어요!
말씀하신 의견에 대한 제 의견도 적어볼게요.

<hr>

> 모든 컴포넌트가 재사용성을 기반으로 구현해야 하는 것인지, 아니면 재사용 가능성이 높다고 기대되는 컴포넌트에 대하여 구현해야 하는 것인지 기준을 잡는다(?)는 게 아직 쉽지 않다고 생각합니다.

1. **미래는 대비하되, 과도하게 대비할 필요는 없다고 생각합니다.** 컴포넌트를 확장해야 하는 사용처가 전혀 없거나, 먼 미래의 일이라 생각되시면 굳이 컴포넌트를 확장할 필요가 없다고 생각합니다.

2. **재사용성이 낮은 컴포넌트는 나쁜 컴포넌트가 아니라고 생각합니다. 목적에 맞지 않는 재사용성을 지닌 컴포넌트가 나쁜 컴포넌트라고 생각합니다.** 컴포넌트에서 재사용성을 챙기면 활용 방안이 무궁무진해지고 중복되는 코드를 줄일 수 있어 좋지만, 그만큼 컴포넌트 내부 자체는 복잡해지고 사용 난이도 또한 증가할 수도 있습니다. 미래를 대비한답시고 쓸 확률이 굉장히 낮은 기능들을 미리 준비해두면 여러모로 관리가 힘드실 수 있을 거라 생각해요. **과도하게 만능인 컴포넌트**요.

3. **재사용성 뿐만 아니라 컴포넌트의 자유도에 따라서도 특징이 달라집니다.** 자유도라는 것은 props로 얼마나 자유로운 값들을 사용할 수 있는지를 의미합니다. 자유도가 높으면 활용도가 높아지나 넓은 옵션으로 인해 개발자가 컴포넌트를 잘못 활용할 가능성도 커질 수 있습니다. 자유도가 낮으면 활용도가 특정 사용처로 집중되며 개발자가 컴포넌트를 잘못 활용할 가능성이 적어지고 일관성 측면에서 좋아지지만, 이후 요구사항이 다양해짐에 따라 이 컴포넌트를 계속해서 수정하거나 담당 개발자를 불러야 하는 단점이 있습니다.

종합적으로 **컴포넌트를 설계하실 때 이 컴포넌트를 사용할 컴포넌트 및 기능을 생각해 보시면 좋을 거라 생각합니다.** 몇 가지 예를 들어보고자 합니다.

- `<Checkbox>`, 단순한 하나의 체크박스

사용 범위 1) - 이 체크박스는 범용성이 상당히 높아. 프로젝트 전체에서 사용할 수 있도록 구현할 거야. 구체적인 디자인은 확정되지 않았지만 다양한 디자인에 사용될 만하니, 프리하게 옵션을 만들어보자. $\rightarrow$ `색상`, `크기` 를 자유롭게 정할 수 있는 props를 추가

사용 범위 2) - 이 체크박스는 범용성이 상당히 높아. 프로젝트 전체에서 사용할 수 있도록 구현할 거야. 근데 우리 프로젝트는 색상이 하나로만 통일된 프로젝트고, 체크박스의 크기도 다양하면 일관성이 떨어질듯해. 사전에 정해놓은 두 사이즈만 쓸 거야. 나는 디자인의 통일성을 위해 범용성을 제한할 거야. $\rightarrow$ `크기`를 제한된 두 옵션으로 선택할 수 있는 props를 추가

- `<RaceGameRankingList>`, 경주 게임 메뉴의 랭킹판 리스트

사용 범위: 이 컴포넌트는 디자인이 단순해 다른 기능에도 충분히 사용될 수 있지만, 나는 오직 부모 컴포넌트인 경주 게임 메뉴를 위해서만 존재하는 하위 컴포넌트로 설계할 거야. 지금의 용도로만 사용되도록 props를 최대한 제한하고, 범용성을 줄이자. $\rightarrow$ 디자인 커스터마이징 관련 props 없이 `사용자 리스트`만을 받는 props 추가

<hr>

> 컴포넌트의 렌더링 이외의 다른 이벤트 리스너로 발생하는 이펙트를 말하는 것으로 알고 있습니다.

렌더링 중에 일어나는 작업들 외부의 작업들을 말씀해 주신 것은 좋지만, 사실 사이드 이펙트는 이벤트 리스너만으로 한정되는 것은 아니라고 생각합니다. 저는 사이드 이펙트를 이렇게 생각합니다. `useEffect` 을 왜 쓰는지, 왜 필요한지를 다루는 미션이니 복잡하게 이야기하지 않을게요.

- 함수 외부의 요소에 영향을 주는 것
- 함수 외부의 요소와 상호작용 하는 것

그렇기 때문에 말씀하신

> 컴포넌트나 함수를 순수하게 작성한다면, 예측 가능성이 올라가면서 랜더링 결과를 예상하기 쉽습니다. 이에 따라 디버깅 및 테스트를 쉽게 할 수 있습니다.

> 렌더링 후에 사이드 이펙트가 일어나야 컴포넌트의 순수함을 지킬 수 있기 때문입니다. 그리고 불필요한 사이드 이펙트의 호출을 막을 수 있다고 생각합니다.

이 중요한 장점들을 유지할 수 있다고 생각합니다. 우리가 사이드 이펙트를 "특별 취급" 하고 `useEffect` 안에 굳이 가둬서 렌더링 프로세스로부터 떼어놓는 이유라 생각해요.

<hr>

궁금하신 점이 추가로 있으시다면 추가로 저한테 말씀해주세요
다음 주차부터는 `self-paced-react-advanced` 라는 한 층 더 응용된 미션을 시작하실 거에요, 지금까지 수행하신 다섯 단계의 미션들은 React의 튼튼한 기초가 되는 것은 물론 React라는 녀석이 작동하는 방식을 눈으로 보며 뭘 챙기려고 한 건지도 엿볼 수 있는 미션이니, 잊지 마시고 이후에도 다시 들러보실 것을 추천드려요

수고 많으셨습니다.
